imports/plugins/core/orders/server/no-meteor/queries/refunds.js

```diff
@@ -0,0 +1,50 @@
+import ReactionError from "@reactioncommerce/reaction-error";
+import { getPaymentMethodConfigByName } from "/imports/plugins/core/payments/server/no-meteor/registration";
+import { getOrderQuery } from "../util/getOrderQuery";
+
+/**
+ * @name refunds
+ * @method
+ * @memberof Order/NoMeteorQueries
+ * @summary Query the Orders collection for an order, and returns refunds applied to payments associated with this order
+ * @param {Object} context - an object containing the per-request state
+ * @param {Object} params - request parameters
+ * @param {String} params.orderId - Order ID
+ * @param {String} params.shopId - Shop ID for the shop that owns the order
+ * @param {String} [params.token] - Anonymous order token
+ * @return {Promise<Array>|undefined} - An array of refunds applied to this order, if found
+ */
+export default async function refunds(context, { orderId, shopId, token } = {}) {
+  const { userHasPermission } = context;
+
+  if (!userHasPermission(["orders", "order/fulfillment", "order/view"], shopId)) throw new ReactionError("access-denied", "User does not have permission");
+
+  if (!orderId || !shopId) {
+    throw new ReactionError("invalid-param", "You must provide orderId and shopId arguments");
+  }
+
+  const selector = getOrderQuery(context, { _id: orderId }, shopId, token);
+  const order = await context.collections.Orders.findOne(selector);
+
+  if (!order) {
+    throw new ReactionError("not-found", "Order not found");
+  }
+
+  const paymentRefunds = [];
+
+  if (Array.isArray(order.payments)) {
+    for (const payment of order.payments) {
+      /* eslint-disable no-await-in-loop */
+      const shopRefunds = await getPaymentMethodConfigByName(payment.name).functions.listRefunds(context, payment);
```

Comment:
Can we make this loop run in parallel by creating the promises without awaiting them then `await Promise.all` after the loop?

Reply:
I think the way I'm doing it has been the pattern we've been using with the non-meteor code. In Meteor code we were using the `Promise.await`, but in all our non-meteor code we are doing the await inside the loop. @aldeed is my assessment correct? is there a reasoning behind why we do it with `await` in the loop more recently?

Reply:
@kieckhafer Actually the pattern is to `await Promise.all` whenever parallel is fine. In many places, parallel is not fine because the order of execution matters or it's transformation chain, and in those instances we `await` in the for loop.

As I typed this it started to sound familiar, and then I remembered this: https://docs.reactioncommerce.com/docs/dev-how-do-i#loop-over-async-function-results

Reply:
In this specific case, it seems to me that parallel would be good.

Reply:
A little confused about how to go about this.

I need to use the results from line 38 on line 40, where would I `await Promise.all`? If I change line 36 to a map, I still need the result from the await inside the map, I can't wait until after it to resolve the promises.

Reply:
Here's what I think I'd do. It's not going to win any beauty pageants but I think it works in parallel and gets you the combination you need of `shopRefunds` and the corresponding `payment` in the same scope. This is from `const paymentRefunds = [];` to the end of the function body.

```js
  const paymentRefunds = [];

  if (Array.isArray(order.payments)) {
    const shopRefundsWithPaymentPromises = order.payments.map((payment) =>
      getPaymentMethodConfigByName(payment.name)
        .functions.listRefunds(context, payment)
        .then((shopRefunds) => shopRefunds, payment));
    // This should give us an arary of objects with properties
    // `shopRefunds` and the corresponding `payment`
    const shopRefundsWithPayments = await Promise.all(shopRefundsWithPaymentPromises);
    shopRefundsWithPayments.forEach(({ shopRefunds, payment }) => {
      const shopRefundsWithPaymentId = shopRefunds.map((shopRefund) => ({
        ...shopRefund,
        paymentId: payment._id,
        paymentDisplayName: payment.displayName
      }));
      paymentRefunds.push(...shopRefundsWithPaymentId);
    });
  }

  return paymentRefunds;

```

Reply:
Got it, this makes sense. I wasn't wrapping my head around that second loop in the `forEach`.